docs/testing/writing-unit-tests.md

```diff
@@ -0,0 +1,456 @@
+    expect(result).toBeTruthy();
+});
+```
+
+Easy enough, just treat it as a common Promise function. This should work with any KBase code that returns a Promise.
+
+***2. Return the Promise***
+
+Another option when working with Promises or similar *thenable* code, is to just return the Promise from the spec and embed the `expect` function in a then block. Something like this:
+
+```Javascript
+it('returns a promise', () => {
+    return codeThatMakesAPromise()
+        .then((result) => {
+            expect(result).toBeDefined();
+        });
+});
+```
+This is similar to the above, but it can also be useful for trapping errors that you want to see (in a `.catch` block instead of `.then`). Though that can also be done with `async/await` in a `try/catch`. 
+
+***3. Backup option - `done` function***
+
+As kind of a last resort, you can make use of the built in `done` function that Jasmine has. This isn't recommended as it's more flaky and can lead to code still running in the background or not completing when you expect. It's mostly useful for code asynchronously that executes but doesn't return a Promise, or relies on callbacks. It's highly recommended to write code that properly uses and returns Promises instead of depending on other techniques, but sometimes there's no other option. In that case, add the `done` variable to any test spec, and it becomes available as a function. Just execute it when your test is done running.
+
+```Javascript
+it('calls "done" when done', (done) => {
+    const myCallback = (result) => {
+        expect(result).toBeTruthy();
+        done();
+    };
+    runCodeThatTakesACallback(myCallback);
+});
+```
+In this test, `runCodeThatTakesACallback` does some kind of code execution, then calls the callback with the result. The only way we can test the result of that code is to write our own callback in the tester that then calls `done` when it's done running.
+
+Examples of where this is needed are jQuery event listeners (e.g. `$x.trigger('someEvent', input)`) or the messages that use the `monoBus` (most messaging that goes through the app cell, input widgets, etc.).
+
+
+### Mocking and Spies
+
+In most cases, you'll want to focus your unit tests on the code you've written. Unit tests shouldn't try to capture flows through every module or KBase service, but should exercise your code's response to input. You're not testing the Workspace service, but whether your viewer can handle workspace data properly.
+
+To that end, many external pieces of code can and should be mocked.
+
+Most of the time, mocks can be made by just setting variables inside functions, or overwriting other functions.
+
+Imagine we have a module like this:
+```Javascript
+define([], () => {
+    function MyModule(dependency) => {
+        function someFunction() => {
+            return dependency.run();
+        }
+
+        return {
+            someFunction: someFunction
+        };
+    }
+
+    return MyModule;
+}
```

Comment:
It might be useful to provide an example that mimics the very common factory pattern:

```js
// GenericKBaseModule.js
define([], () => {
    function factory(args) {
        const config = args;

        function start(arg) {
            return Promise.try(() => {
                // whatever
            });
        }

        function stop() {
            return Promise.try(() => {
                // whatever
            });
        }

        return {
            start: start,
            stop: stop,
        };
    }

    return {
        make: function (args) {
            return factory(args);
        },
    };
```

and then give examples of testing `GenericKBaseModule.make()`, and the `start` and `stop` functions of the object returned by `make`. Part of the annoyingness of testing the KBase codebase is trying to work around the limitations of only being able to test external effects of object methods, and not having access to internal methods or variables.

Reply:
this could go at the bottom of this segment as you'd probably want demonstrate testing `start` and `stop` asynchronously and/or using spies.